src/burnchains/db.rs

```diff
@@ -147,16 +262,554 @@ impl<'a> BurnchainDBTransaction<'a> {
+                    test_debug!(
+                        "No block-commit parent found for {},{},{}",
+                        &commit.txid,
+                        commit.block_height,
+                        commit.vtxindex
+                    );
+                }
+                self.update_block_commit_affirmation(commit, None, 0)
+                    .map_err(|e| BurnchainError::from(e))?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Create a prepare-phase affirmation map.  This is only done at the very end of a reward
+    /// cycle, once the anchor block is chosen and a new reward cycle is about to begin.  This
+    /// method updates the prepare-phase block-commit's affirmation map to reflect what its miner
+    /// believes to be the state of all anchor blocks, _including_ this new reward cycle's anchor
+    /// block.
+    pub fn make_prepare_phase_affirmation_map<B: BurnchainHeaderReader>(
         &self,
-        block_hash: &BurnchainHeaderHash,
+        indexer: &B,
+        burnchain: &Burnchain,
+        reward_cycle: u64,
+        block_commit: &LeaderBlockCommitOp,
+        anchor_block: Option<&LeaderBlockCommitOp>,
+        descends_from_anchor_block: bool,
+    ) -> Result<u64, BurnchainError> {
+        test_debug!(
+            "Make affirmation map for {},{},{} (parent {},{}) in reward cycle {}",
+            &block_commit.txid,
+            block_commit.block_height,
+            block_commit.vtxindex,
+            block_commit.parent_block_ptr,
+            block_commit.parent_vtxindex,
+            reward_cycle
+        );
+
+        let parent = match BurnchainDB::get_commit_at(
+            &self.sql_tx,
+            indexer,
+            block_commit.parent_block_ptr,
+            block_commit.parent_vtxindex,
+        )? {
+            Some(p) => p,
+            None => {
+                if block_commit.parent_block_ptr == 0 && block_commit.vtxindex == 0 {
+                    test_debug!(
+                        "Prepare-phase commit {},{},{} builds off of genesis",
+                        &block_commit.block_header_hash,
+                        block_commit.block_height,
+                        block_commit.vtxindex
+                    );
+                } else {
+                    test_debug!(
+                        "Prepare-phase commit {},{},{} has no parent, so must be invalid",
+                        &block_commit.block_header_hash,
+                        block_commit.block_height,
+                        block_commit.vtxindex
+                    );
+                }
+                return Ok(0);
+            }
+        };
+
+        let parent_metadata =
+            BurnchainDB::get_commit_metadata(&self.sql_tx, &parent.burn_header_hash, &parent.txid)?
+                .expect("BUG: no metadata found for parent block-commit");
+
+        let (am, affirmed_reward_cycle) = if let Some(ab) = anchor_block {
```

Comment:
Nit: I know you prefer not to shadow matched options -- but I have to argue that a shadowed:

```
if let Some(anchor_block) = anchor_block {
...
```

Is more clear than introducing the `ab` variable.

Reply:
Done